From: hololeap <hololeap@protonmail.com>
Signed-off-by: hololeap <hololeap@protonmail.com>
Bug: https://github.com/gentoo-haskell/gentoo-haskell/issues/1231
Bug: https://github.com/gentoo-haskell/gentoo-haskell/issues/1270

doctest pulls in the out-of-scope packages if they are installed on the
system. The best workaround is to use migrate to cabal-doctest.

The temporary package database at "${S}/package.conf.d" is used to avoid GHC
panics.

diff -urN generic-lens-2.2.1.0/Setup.hs generic-lens-2.2.1.0-r1/Setup.hs
--- generic-lens-2.2.1.0/Setup.hs	2001-09-08 19:46:40.000000000 -0600
+++ generic-lens-2.2.1.0-r1/Setup.hs	2022-03-31 13:32:46.128447775 -0600
@@ -1,2 +1,6 @@
-import Distribution.Simple
-main = defaultMain
+module Main (main) where
+
+import Distribution.Extra.Doctest ( defaultMainWithDoctests )
+
+main :: IO ()
+main = defaultMainWithDoctests "doctests"
diff -urN generic-lens-2.2.1.0/examples/doctest.hs generic-lens-2.2.1.0-r1/examples/doctest.hs
--- generic-lens-2.2.1.0/examples/doctest.hs	2001-09-08 19:46:40.000000000 -0600
+++ generic-lens-2.2.1.0-r1/examples/doctest.hs	2022-03-31 15:38:38.738358107 -0600
@@ -1,6 +1,31 @@
 import Test.DocTest
-main
-  = doctest
-      [ "src"
-      , "examples"
-      ]
+
+import GHC.IO.Encoding (setLocaleEncoding)
+import System.Directory (getCurrentDirectory, makeAbsolute)
+import System.FilePath.Posix ((</>))
+import System.IO (utf8)
+import Build_doctests (pkgs)
+ 
+main :: IO ()
+main = do
+    setLocaleEncoding utf8
+    pwd    <- getCurrentDirectory
+    prefix <- makeAbsolute pwd
+
+    let customFlags =
+            [ "-package-env=-"
+            , "-hide-all-packages"
+            , "-no-global-package-db"
+            , "-no-user-package-db"
+            , "-package-db=" ++ prefix </> "dist/package.conf.inplace"
+            , "-package-db=" ++ prefix </> "package.conf.d"
+            ]
+
+    let modules =
+            [ "src"
+            , "examples"
+            ]
+
+    let args = customFlags ++ pkgs ++ modules
+
+    doctest args
diff -urN generic-lens-2.2.1.0/generic-lens.cabal generic-lens-2.2.1.0-r1/generic-lens.cabal
--- generic-lens-2.2.1.0/generic-lens.cabal	2022-03-31 13:20:12.958456717 -0600
+++ generic-lens-2.2.1.0-r1/generic-lens.cabal	2022-03-31 14:52:42.858390826 -0600
@@ -12,7 +12,7 @@
 author:               Csongor Kiss
 maintainer:           kiss.csongor.kiss@gmail.com
 category:             Generics, Records, Lens
-build-type:           Simple
+build-type:           Custom
 cabal-version:        >= 1.10
 Tested-With:          GHC == 8.4.1, GHC == 8.4.4, GHC == 8.6.5, GHC == 8.8.4, GHC == 8.10.3, GHC == 9.0.1, GHC == 9.2.1
 
@@ -20,6 +20,12 @@
                     , examples/Examples.hs
                     , ChangeLog.md
 
+custom-setup
+    setup-depends:
+        base
+      , Cabal
+      , cabal-doctest
+
 library
   exposed-modules:    Data.Generics.Wrapped
                     , Data.Generics.Product
@@ -102,4 +108,8 @@
   ghc-options:        -threaded
   main-is:            doctest.hs
   build-depends:      base >4 && <5, doctest
+                    , directory
+                    , filepath
+                    , base-compat
+                    , lens
   hs-source-dirs:     examples
